File: ruby/spec/03_stacks_and_queues/02_stack_min.rb
# n - stack size
# O(1) for push, pop and min methods
class StackWithMin
  class SNode
    attr_reader :value, :prev, :min

    def initialize(value, prev_min, prev = nil)
      @value = value
      @prev = prev
      @min = value < prev_min ? value : prev_min
    end
  end

  def initialize
    @tail = nil
  end

  def push(value)
    if @tail
      @tail = SNode.new(value, @tail.min, @tail)
    else
      @tail = SNode.new(value, value)
    end
  end

  def pop
    if @tail
      value = @tail.value
      @tail = @tail.prev
      value
    end
  end

  def min
    @tail&.min
  end

  def to_s
    arr = []
    curr = @tail

    while curr do
      arr << "v=#{curr.value} m=#{curr.min}"
      curr = curr.prev
    end

    '|[' + arr.reverse.map(&:to_s).join(', ') + ']'
  end
end

RSpec.describe StackWithMin do
  let(:stack) { described_class.new }

  describe 'push - min - pop' do
    it do
      stack.push(1)
      expect(stack.min).to eq 1

      stack.push(2)
      expect(stack.min).to eq 1

      stack.push(0)
      expect(stack.min).to eq 0

      stack.push(1)
      expect(stack.min).to eq 0

      expect(stack.pop).to eq 1
      expect(stack.min).to eq 0

      expect(stack.pop).to eq 0
      expect(stack.min).to eq 1

      expect(stack.pop).to eq 2
      expect(stack.min).to eq 1

      expect(stack.pop).to eq 1
      expect(stack.min).to eq nil

      expect(stack.pop).to eq nil
    end
  end
end

File: go/03_stacks_and_queues/02_stack_min.go
package stacks_and_queues

import (
	"bytes"
	"errors"
	"fmt"
)

// n - stack size
// O(1) for push, pop and min methods
type stackWithMin struct {
	tail *sNode
}

type sNode struct {
	value int
	min   int
	prev  *sNode
}

func NewSNode(value int, prev_min int, prev *sNode) *sNode {
	node := &sNode{value: value, prev: prev}

	if value < prev_min {
		node.min = value
	} else {
		node.min = prev_min
	}

	return node
}

func (s *stackWithMin) Push(value int) *stackWithMin {
	if s.tail != nil {
		s.tail = NewSNode(value, s.tail.min, s.tail)
	} else {
		s.tail = NewSNode(value, value, nil)
	}

	return s
}

func (s *stackWithMin) Pop() (int, error) {
	if s.tail != nil {
		value := s.tail.value
		s.tail = s.tail.prev
		return value, nil
	} else {
		return -1, errors.New("Stack is empty")
	}
}

func (s *stackWithMin) Min() (int, error) {
	if s.tail != nil {
		return s.tail.min, nil
	} else {
		return -1, errors.New("Stack is empty")
	}
}

func (s *stackWithMin) ToString() string {
	arr := []string{}
	curr := s.tail

	for curr != nil {
		arr = append(arr, fmt.Sprintf("v=%d m=%d", curr.value, curr.min))
		curr = curr.prev
	}

	res := bytes.Buffer{}
	res.WriteString("||")

	for i, s := range ReverseStrings(arr) {
		res.WriteString(s)

		if i < len(arr)-1 {
			res.WriteString(", ")
		}
	}

	res.WriteString("]")
	return res.String()
}
